grudge/op.py

```diff
@@ -477,14 +477,35 @@ def _cross_rank_trace_pairs_scalar_field(dcoll, vec, tag=None):
         return [rbcomm.finish() for rbcomm in rbcomms]
 
 
-def cross_rank_trace_pairs(dcoll, vec, tag=None):
-    if isinstance(vec, np.ndarray):
+def cross_rank_trace_pairs(dcoll, ary, tag=None):
+    r"""Get a list of *ary* trace pairs for each partition boundary.
+
+    For each partition boundary, the field data values in *ary* are
+    communicated to/from the neighboring partition. Presumably, this
+    communication is MPI (but strictly speaking, may not be, and this
+    routine is agnostic to the underlying communication, see e.g.
+    _cross_rank_trace_pairs_scalar_field).
+
+    For each face on each partition boundary, a :class:`TracePair` is
+    created with the locally, and remotely owned partition boundary face
+    data as the `internal`, and `external` components, respectively.
+    Each of the TracePair components are structured like *ary*.
+
+    The input field data *ary* may be a single
+    :class:`~meshmode.dof_array.DOFArray`, or an object
+    array of ``DOFArray``\ s of arbitrary shape.
+    """
+    if isinstance(ary, np.ndarray):
+        oshape = ary.shape
+        comm_vec = ary.flatten()
 
-        n, = vec.shape
+        n, = comm_vec.shape
         result = {}
+        # FIXME: Batch this communication rather than
+        # doing it in sequence.
```

Comment:
Just note to selves; it may be worthwhile leaving the option to chunk into the smallest (reasonably) possible message.

Reply:
I thought "bigger is better" when it comes to MPI comm?

Reply:
Me too, but my experience is based entirely on CPU-type experience.  I thought we had heard recently from Shelby that message latency was getting quite high for even modestly sized messages, potentially favoring smaller messages. I could be mistaken about that 🤷‍♂️

Reply:
True... but I'd regard all those cases as bugs in the MPI implementation. @sll2, would you agree?

Reply:
Even if we keep sending separate messages for each component, I imagine it would be preferable to fire them all off and then wait once instead of waiting on each component sequentially, right?

Reply:
> Even if we keep sending separate messages for each component, I imagine it would be preferable to fire them all off and then wait once instead of waiting on each component sequentially, right?

Yep, for sure.

Reply:
> Even if we keep sending separate messages for each component, I imagine it would be preferable to fire them all off and then wait once instead of waiting on each component sequentially, right?

As opposed to repeated calls to something like (send/recv wait, send/recv wait, ... ), definitely would be better to do (send/recv, send/recv, wait), yeah. Unless we bump up against some data transfer resource limitation due to how much data we have in simultaneous flight.

Reply:
When to batch communication is motivated by the dominating term in the Max-Rate Model. If we're already incurring the injection rate bandwidth of the NIC, then the difference in time required for every process to send 10^6 Bytes of data vs every process sending double that amount appears to be small. 

More information:
When the entire amount of data being injected by a node is large (~10^5 Bytes or higher on Lassen), it's faster to send the data when it's split up evenly across all active processes on a node. This is fairly consistent across machines (Blue Waters, Lassen, Summit) and MPI implementations (spectrum-mpi, mvapich2), so it's not really an artifact of the MPI implementation.

Here's an example below for Lassen, where the Message size is being split up across various processes per node, and the fastest time is achieved by splitting that data sent evenly across all available cores (PPN 40).
[lassen_cpu_node_pong.pdf](https://github.com/inducer/grudge/files/6336983/lassen_cpu_node_pong.pdf)

In our case, however, presumably the data being sent off node is already split evenly across the active processes.

Reply:
Just for context...

At each RK substep of our production problem, we need to exchange ~500Bytes total per partition boundary (PB) point.  This includes exchanging only **Q** and grad(**Q**) - safe for now.

That translates to about (2KB, 3KB, 5KB) per boundary triangle for order (1, 2, 3) triangles, respectively.  The lower bound (i.e. the (reasonably) smallest) message size is that where we bust it into one message per scalar component so ~(24B, 50B, 100B) per boundary triangle.

The total message size depends on the number of triangles on each PB, which is clearly a function of mesh resolution and partitioning. For say 100 order 1 triangles on each PB that means the smallest possible message size for order 1 runs is ~3KB, and big message of ~20KB for a (fully) batched exchange.